nixos/modules/system/activation/switch-to-configuration.py

```diff
@@ -0,0 +1,604 @@
+oldVersion = read_if_exists("/run/current-system/init-interface-version", otherwise="")
+newVersion = open(os.path.join(out, "init-interface-version")).read()
+
+if newVersion != oldVersion:
+    print("""
+        Warning: the new NixOS configuration has an ‘init’ that is
+        incompatible with the current configuration.  The new configuration
+        won\'t take effect until you reboot the system.
+    """, file=sys.stderr)
+    sys.exit(100)
+
+# Ignore SIGHUP so that we're not killed if we're running on (say)
+# virtual console 1 and we restart the "tty1" unit.
+import signal
+signal.signal(signal.SIGPIPE, signal.SIG_IGN)
+
+def get_active_units():
+    import dbus
+
+    bus = dbus.SystemBus()
+    obj = bus.get_object('org.freedesktop.systemd1',
+                         '/org/freedesktop/systemd1')
+
+    units = obj.ListUnitsByPatterns([], [],
+                                    dbus_interface='org.freedesktop.systemd1.Manager')
+    res = {}
+    for unit in units:
+        (id, description, load_state, active_state, sub_state,
+            following, unit_path, job_id, job_type, job_path) = unit
+        if following != '':
+            continue
+        if job_id == 0 and active_state == 'inactive':
+            continue
+        res[id] = {
+            'load': load_state,
+            'state': active_state,
+            'substate': sub_state
+        }
+
+    return res
+
+def parse_fstab(filename):
+    import re
+    fss = []
+    swaps = []
+    for line in open(filename, 'r').readlines():
+        line = line.strip()
+        line = re.sub(r'^\s*#.*', '', line)
+        if line.strip() == '':
+            continue
+
+        xs = re.split(' ', line)
+        device = xs[0]
+        mountpoint = xs[1]
+        fs_type = xs[2]
+        options = xs[3] if len(xs) > 4 else ''
+        if fs_type == 'swap':
+            swaps[device] = { 'options': options }
+        else:
+            fss[mountpoint] = {
+                'device': device,
+                'fs_type': fs_type,
+                'options': options
+            }
+
+    return (fss, swaps)
+
+def parse_unit(filename):
+    info = {}
+    info.update(parse_key_values(read_if_exists(filename)))
+    overrides = '%s.d/overrides.conf' % filename
+    info.update(parse_key_values(read_if_exists(overrides)))
+    return info
+
+def parse_key_values(s):
+    import re
+    info = {}
+    for line in s.split('\n'):
+        m = re.match(r'^([^=]+)=(.*)$', line)
+        if m is None:
+            die('error parsing unit')
+        info[m.group(1)] = m.group(2)
+
+    return info
+
+def bool_is_true(s: str) -> bool:
+    return s in ['yes', 'true']
+
+# As a fingerprint for determining whether a unit has changed, we use
+# its absolute path. If it has an override file, we append *its*
+# absolute path as well.
+def fingerprint_unit(filename):
+    fprint = os.path.abspath(filename)
+    override = '%s.d/overrides.conf' % filename
```

Comment:
use .format

Reply:
or python3.6 format strings

Reply:
Are we ready to accept a >= Python 3.6 as a requirement for NixOS?

Reply:
Yes. python36 is the default python3, and soon it will be python37.